src/renderer/contexts/BackendContext.tsx

```diff
@@ -37,27 +46,130 @@ export const BackendContext = createContext<Readonly<BackendContextState>>(
 
 interface BackendProviderProps {
     url: string;
-    schemata: SchemaMap;
     pythonInfo: PythonInfo;
+}
+
+interface NodesInfo {
+    rawResponse: BackendNodesResponse;
+    schemata: SchemaMap;
     categories: Category[];
-    categoriesMissingNodes: string[];
     functionDefinitions: Map<SchemaId, FunctionDefinition>;
-    refreshNodes: () => void;
+    categoriesMissingNodes: string[];
 }
 
-export const BackendProvider = memo(
-    ({
-        url,
-        schemata,
-        pythonInfo,
+const processBackendResponse = (rawResponse: BackendNodesResponse): NodesInfo => {
+    const { categories, categoriesMissingNodes, nodes } = rawResponse;
+
+    return {
+        rawResponse,
+        schemata: new SchemaMap(nodes),
         categories,
+        functionDefinitions: parseFunctionDefinitions(nodes),
         categoriesMissingNodes,
-        functionDefinitions,
-        refreshNodes,
-        children,
-    }: React.PropsWithChildren<BackendProviderProps>) => {
+    };
+};
+
+export const BackendProvider = memo(
+    ({ url, pythonInfo, children }: React.PropsWithChildren<BackendProviderProps>) => {
+        const { t } = useTranslation();
         const backend = getBackend(url);
 
+        const { sendAlert, forgetAlert } = useContext(AlertBoxContext);
```

Comment:
I'm not a fan of all this code just existing here. Could we extract all the code you inserted out into a function (hook)? I think it only needs to return `nodesInfo` and the current connection state. Feel free to keep it in this file, I just don't want to make `BackendProvider` this complex.